libr/bin/format/elf/elf.c

```diff
@@ -2627,150 +2642,163 @@ char *Elf_(r_bin_elf_get_rpath)(ELFOBJ *bin) {
-		}
-		for (j = 0; j < bin->g_sections[i].size; j += res) {
-			if (bin->g_sections[i].size > bin->size) {
-				break;
-			}
-			if (bin->g_sections[i].offset > bin->size) {
-				break;
-			}
-			if (rel >= reloc_num) {
-				bprintf ("Internal error: ELF relocation buffer too small,"
-				         "please file a bug report.");
-				break;
-			}
-			if (!bin->is_rela) {
-				rela = is_rela? DT_RELA : DT_REL;
-			} else {
-				rela = bin->is_rela;
-			}
-			res = read_reloc (bin, &ret[rel], rela, bin->g_sections[i].offset + j);
-			if (j + res > bin->g_sections[i].size) {
-				bprintf ("malformed file, relocation entry #%u is partially beyond the end of section %u.\n", rel, i);
-			}
-			if (is_bin_etrel (bin)) {
-				if (bin->g_sections[i].info < bin->ehdr.e_shnum && bin->shdr) {
-					ret[rel].rva = bin->shdr[bin->g_sections[i].info].sh_offset + ret[rel].offset;
-					ret[rel].rva = Elf_(r_bin_elf_p2v) (bin, ret[rel].rva);
-				} else {
-					ret[rel].rva = ret[rel].offset;
-				}
-			} else {
-				ret[rel].rva = ret[rel].offset;
-				ret[rel].offset = Elf_(r_bin_elf_v2p) (bin, ret[rel].offset);
-			}
-			ret[rel].last = 0;
-			if (res < 0) {
-				break;
-			}
-			rel++;
+}
+
+static size_t get_num_relocs(struct dynamic_relocation_section *info) {
+	size_t res = 0;
+
+	if (info->relaent) {
+		res += info->rela_size / info->relaent;
+	}
+
+	if (info->relent) {
+		res += info->rel_size / info->relent;
+	}
+
+	if (info->jmprel_size) {
+		size_t size = info->plt_mode == DT_RELA? sizeof (Elf_(Rela)): sizeof (Elf_(Rel));
+		res += info->jmprel_size / size;
+	}
+
+	return res;
+}
+
+static struct dynamic_relocation_section *get_dynamic_info(ELFOBJ *bin) {
+	struct dynamic_relocation_section *res = calloc (1, sizeof (struct dynamic_relocation_section));
+
+	for (size_t i = 0; i < bin->dyn_entries; ++i) {
+		Elf_(Dyn) *dyn_struct = bin->dyn_buf + i;
+
+		switch (dyn_struct->d_tag) {
+		case DT_JMPREL:
+			res->addr_jmprel = dyn_struct->d_un.d_ptr;
+			break;
+		case DT_PLTRELSZ:
+			res->jmprel_size = dyn_struct->d_un.d_val;
+			break;
+		case DT_PLTREL:
+			res->plt_mode = dyn_struct->d_un.d_val;
+			break;
+		case DT_RELA:
+			res->addr_rela = dyn_struct->d_un.d_ptr;
+			break;
+		case DT_RELASZ:
+			res->rela_size = dyn_struct->d_un.d_val;
+			break;
+		case DT_RELAENT:
+			res->relaent = dyn_struct->d_un.d_val;
+			break;
+		case DT_REL:
+			res->addr_rel = dyn_struct->d_un.d_ptr;
+			break;
+		case DT_RELSZ:
+			res->rel_size = dyn_struct->d_un.d_val;
+			break;
+		case DT_RELENT:
+			res->relent = dyn_struct->d_un.d_val;
+			break;
 		}
 	}
-	ret[reloc_num].last = 1;
+
+	return res;
+}
+
+RBinElfReloc* Elf_(r_bin_elf_get_relocs)(ELFOBJ *bin) {
+	RBinElfReloc *ret = NULL;
+
+	if (!bin || !bin->g_sections) {
+		return NULL;
+	}
+
+	struct dynamic_relocation_section *info = get_dynamic_info (bin);
+	size_t num_relocs = get_num_relocs (info);
+	ret = calloc (num_relocs + 1, sizeof (RBinElfReloc));
+	populate_relocs_record (bin, ret, info);
```

Comment:
better do something like:
```C
ret = populate_relocs_record (bin, info);
```

and let `populate_relocs_record` allocate the memory.